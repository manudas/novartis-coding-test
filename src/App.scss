.App {
  text-align: center;

  &-filters {
    display: flex;
    flex-direction: row;
    font-size: calc(10px + 2vmin);

    padding: 10px;
    justify-content: space-evenly;
  }

  &-filter {
    flex-grow: 0.3;
  }
}
