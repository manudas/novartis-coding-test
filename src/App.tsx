import {
  useEffect,
  useState,
} from 'react'

import { Filters } from 'components/filters'
import { HistogramGraph } from 'components/histogramGraph'

import {
  fetchData,
  getMapKey
} from 'helpers'

import {
  FETCHED_DATA,
  SampleDef,
  CcleExpressionDef,
  CACHED_DATA,
  GeneSymbolDef,
  LineageDef
} from 'types'

import './App.scss'

function App() {

  const [ data, setCachedData ] = useState<Map<string, CACHED_DATA[]>>(new Map())
  const [ geneSymbol, setGeneSymbol ] = useState<GeneSymbolDef | undefined>()
  const [ lineage, setLineage ] = useState<LineageDef | undefined>()
  const [ KRASGeneDescription, setKRASGeneDescription ] = useState<string | undefined>()
  const [ TP53GeneDescription, setTP53GeneDescription ] = useState<string | undefined>()

  /**
   * Builds, sorts and caches data from Samples
   * and CCLE Expressions fetched from the API
   *
   * @param data {FETCHED_DATA} data
   */
  const buildCache = (data: FETCHED_DATA): void => {
    const {
      samples: samples_data,
      ccle_expression: expresions_data,
    } = data;

    const samples_by_DepMap_ID = samples_data.reduce((map, sample) => {
      map[sample.DepMap_ID] = sample
      return map
    }, {} as {[id: string]: SampleDef})

    const expressions_by_DepMap_ID = expresions_data.reduce((map, expression) => {
      if (!map[expression.DepMap_ID]) map[expression.DepMap_ID] = []
      map[expression.DepMap_ID].push(expression)
      return map
    }, {} as {[id: string]: CcleExpressionDef[]})

    // Let's use a set to remove duplicates and concatenate keys in case
    // we have samples and not corresponding expressions or vice-versa
    const DepMap_ID_set = new Set(Object.keys(samples_by_DepMap_ID).concat(Object.keys(expressions_by_DepMap_ID)))

    const cache = new Map<string, CACHED_DATA[]>()
    DepMap_ID_set.forEach(id => {
      const {
        lineage,
        lineage_subtype
      } = samples_by_DepMap_ID[id]
      const expressions = expressions_by_DepMap_ID[id]

      expressions && expressions.forEach(({
        gene_symbol,
        expression_value,
      }) => {

        const mapKey = getMapKey(gene_symbol, lineage)

        /*
          CACHED_DATA object shape:

          DepMap_ID: string
          lineage: 'blood' | 'lung'
          lineage_subtype: string | null
          gene_symbol: 'TP53' | 'KRAS'
          expression_value: number

        */
        const arr_data: CACHED_DATA[] = cache.get(mapKey) ?? []
        arr_data.push({
          DepMap_ID: id,
          lineage,
          lineage_subtype,
          gene_symbol,
          expression_value
        })
        cache.set(mapKey, arr_data)
      })

      // Sorting cache in descending order
      cache.forEach(arr_data => {
        arr_data.sort((a, b) => b.expression_value - a.expression_value)
      })

      // Setting the cached data to the state
      setCachedData(cache)
    })

  }

  useEffect(() => {
    const getData = async () => {
      const data: any = await Promise.allSettled([
        fetchData(),
        (await fetch('https://rest.ensembl.org/lookup/id/ENSG00000133703?content-type=application/json')).json(),
        (await fetch('https://rest.ensembl.org/lookup/id/ENSG00000141510?content-type=application/json')).json(),
      ])

      const [{
        value: sampleExpressionData
      }, {
        value: {
          description: KRAS_GeneDescription = undefined
        } = {}
      }, {
        value: {
          description: TP53_GeneDescription = undefined
        } = {}
      }] = data

      setKRASGeneDescription(KRAS_GeneDescription)
      setTP53GeneDescription(TP53_GeneDescription)
      buildCache(sampleExpressionData)
    }

    getData()
  }, [])

  return (
    <div className="App">
      <Filters
        geneSymbol={geneSymbol}
        lineage={lineage}
        changeGeneSymbolHandler={({
          target: {
            value
          }
        }) => {
          setGeneSymbol(value as GeneSymbolDef)
        }}
        changeLineageHandler={({
          target: {
            value
          }
        }) => {
          setLineage(value as LineageDef)
        }}
      />
      <HistogramGraph
        geneSymbol={geneSymbol}
        lineage={lineage}
        data={data}
        geneDescriptions={{
          KRAS: KRASGeneDescription,
          TP53: TP53GeneDescription
        }}
      />
    </div>
  );
}

export default App
