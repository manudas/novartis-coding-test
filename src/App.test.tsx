import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';

describe('Test suite for Novartis Bioinformatis Code challenge APP', () => {

  test('renders the app and check the presence of the', () => {
    const { container } = render(<App />)
    const filterGene = screen.getAllByText(/gene symbol/i)
    expect(filterGene[0]).toBeInTheDocument()
    const lineage = screen.getAllByText(/lineage/i)
    expect(lineage[0]).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  });
})