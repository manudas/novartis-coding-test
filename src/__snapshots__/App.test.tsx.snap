// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for Novartis Bioinformatis Code challenge APP renders the app and check the presence of the 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-filters"
    >
      <div
        class="MuiFormControl-root App-filter css-1nrlq1o-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="gene-symbol-label"
        >
          Gene Symbol
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="gene-symbol-label gene-symbol"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="gene-symbol"
            role="button"
            tabindex="0"
          >
            -
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="-"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Gene Symbol
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root App-filter css-1nrlq1o-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="lineage-label"
        >
          Lineage
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="lineage-label lineage"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="lineage"
            role="button"
            tabindex="0"
          >
            -
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="-"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Lineage
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </header>
    <div
      style="position: relative; display: inline-block;"
    />
  </div>
</div>
`;
